### ISSUE TITLE
OneCycleLearningrate is bug in version

### ISSUE THREAD

**Issue by tinhchuquang** [pytorch/pytorch#32898]
## 🐛 Bug

```
 scheduler.step()       
 File "anaconda3/lib/python3.6/site-packages/torch/optim/lr_scheduler.py", line 150, in step
for param_group, lr in zip(self.optimizer.param_groups,values):
UnboundLocalError: local variable 'values' referenced before assignment
```
## To Reproduce

1. `scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer, 1e-2, total_steps=None, epochs=n_epochs, steps_per_epoch=5021, pct_start=0.0,anneal_strategy='cos', cycle_momentum=True, base_momentum=0.85, max_momentum=0.95, div_factor=100.0)`
2. `optimizer.step()`
    `optimizer.zero_grad()`
3. `scheduler.step()`

## Environment
 - PyTorch Version (e.g., 1.4): 1.4.0
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source): pip 
 - Build command you used (if compiling from source):  pip install torch
 - Python version: 3.6
 - CUDA/cuDNN version: 9.2
 - GPU models and configuration: no config
 - Any other relevant information:

cc @ezyang @gchanan @zou3519 @vincentqb @SsnL @albanD @gqchen

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/32639

**Comment by tinhchuquang** [comment]
Thanks @vadimkantorov

**Comment by vadimkantorov** [comment]
I think you should keep this issue open / reopen it. My original issue was closed without proper investigation (instead there happened architectural discussion)

**Comment by ezyang** [comment]
@smessmer for future reference, this is an optimizer issue, not an autograd issue

**Comment by ezyang** [comment]
This seems like a major bug, since the code is "obviously" wrong

**Comment by vincentqb** [comment]
Can you provide a full code snippet to reproduce the error? I am unable to do so here.

```
import torch 

print(torch.__version__)  # 1.4.0
from torch.nn import Parameter 
from torch.optim import SGD 
from torch.optim.lr_scheduler import OneCycleLR 

model = [Parameter(torch.randn(2, 2, requires_grad=True))] 
optimizer = SGD(model, 0.1) 

scheduler = OneCycleLR(optimizer, 1e-2, total_steps=None, epochs=10, steps_per_epoch=5021, pct_start=0.0,anneal_strategy='cos', cycle_momentum=True, base_momentum=0.85, max_momentum=0.95, div_factor=100.0) 

optimizer.step() 
optimizer.zero_grad() 
scheduler.step()   
```

That being said, it does seem related to #32639. In this case, the `get_lr` function was overridden with a function taking an `epoch` parameter, which is not allowed. However, the relevant error was not being raised. #32997 addresses this by letting the relevant error being raised.

**Comment by tinhchuquang** [comment]
@vincentqb I Try with no data generator and no error
```
import torch

print(torch.__version__)  # 1.4.0
from optimizer_code import Over9000
from model import EfficientNet

model = EfficientNet.from_pretrained('efficientnet-b0')
optimizer = Over9000(model.parameters(), lr=2e-3, weight_decay=1e-3)

scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer, 1e-2, total_steps=None, epochs=1,
                                                steps_per_epoch=5021, pct_start=0.0, anneal_strategy='cos',
                                                cycle_momentum=True, base_momentum=0.85, max_momentum=0.95,
                                                div_factor=100.0)

optimizer.step()
optimizer.zero_grad()
scheduler.step()
```
You try with data generator will error  "UnboundLocalError: local variable 'values' referenced before assignment"

**Comment by tinhchuquang** [comment]
After 4 epoch, will error

![Screenshot from 2020-02-06 08-37-07](https://user-images.githubusercontent.com/21126182/73898515-32b51080-48bc-11ea-86b0-d6289c5d776a.png)

**Comment by tinhchuquang** [comment]
```
import argparse

import torch
import torch.nn as nn
import torch.nn.functional as F
from torchvision import datasets, transforms

from optimizer_code import Over9000

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout2d(0.25)
        self.dropout2 = nn.Dropout2d(0.5)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output


def train(args, model, device, train_loader, optimizer, epoch, scheduler):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        scheduler.step()
        if batch_idx % args.log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                       100. * batch_idx / len(train_loader), loss.item()))


def test(args, model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))


def main():
    # Training settings
    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
    parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                        help='input batch size for training (default: 64)')
    parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                        help='input batch size for testing (default: 1000)')
    parser.add_argument('--epochs', type=int, default=14, metavar='N',
                        help='number of epochs to train (default: 14)')
    parser.add_argument('--lr', type=float, default=1.0, metavar='LR',
                        help='learning rate (default: 1.0)')
    parser.add_argument('--gamma', type=float, default=0.7, metavar='M',
                        help='Learning rate step gamma (default: 0.7)')
    parser.add_argument('--no-cuda', action='store_true', default=False,
                        help='disables CUDA training')
    parser.add_argument('--seed', type=int, default=1, metavar='S',
                        help='random seed (default: 1)')
    parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                        help='how many batches to wait before logging training status')

    parser.add_argument('--save-model', action='store_true', default=False,
                        help='For Saving the current Model')
    args = parser.parse_args()
    use_cuda = not args.no_cuda and torch.cuda.is_available()

    torch.manual_seed(args.seed)

    device = torch.device("cuda" if use_cuda else "cpu")

    kwargs = {'num_workers': 1, 'pin_memory': True} if use_cuda else {}
    train_loader = torch.utils.data.DataLoader(
        datasets.MNIST('../data', train=True, download=True,
                       transform=transforms.Compose([
                           transforms.ToTensor(),
                           transforms.Normalize((0.1307,), (0.3081,))
                       ])),
        batch_size=args.batch_size, shuffle=True, **kwargs)
    test_loader = torch.utils.data.DataLoader(
        datasets.MNIST('../data', train=False, transform=transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.1307,), (0.3081,))
        ])),
        batch_size=args.test_batch_size, shuffle=True, **kwargs)

    model = Net().to(device)

    optimizer = Over9000(model.parameters(), lr=2e-3, weight_decay=1e-3)

    scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer, 1e-2, total_steps=None, epochs=1,
                                                    steps_per_epoch=5021, pct_start=0.0, anneal_strategy='cos',
                                                    cycle_momentum=True, base_momentum=0.85, max_momentum=0.95,
                                                    div_factor=100.0)
    for epoch in range(1, args.epochs + 1):
        train(args, model, device, train_loader, optimizer, epoch, scheduler)
        test(args, model, device, test_loader)
        # scheduler.step()

    if args.save_model:
        torch.save(model.state_dict(), "mnist_cnn.pt")


if __name__ == '__main__':
    main()
```

Error after 6 epoch.

**Comment by fmassa** [comment]
@vincentqb says that the fix should be landing soon.

**Comment by vincentqb** [comment]
I cannot reproduce your particular error. Can you provide a minimal example using only pytorch command? I do not have ` Over9000` in `optimizer_code`, nor  `EfficientNet` in `model`,
from [comment](https://github.com/pytorch/pytorch/issues/32898#issuecomment-582275214)

Also, #32997 has landed which address this general error you are seeing. Can you try again by compiling from source?

**Comment by rsk2327** [comment]
@vincentqb The error arises when you call `scheduler.step` more number of times than the max number of steps specified when instantiating the scheduler. So if the scheduler is planned to take 800 steps, then on the 801th scheduler.step() call, it returns the above error.

I can share code to reproduce the error if required.

**Comment by vincentqb** [comment]
@rsk2327 -- have you tried with the latest code on master? (in particular, after #32997)

**Comment by Voldemorten** [comment]
I too am experiencing this issue, but in quite a different way. 
Minimal code to reproduce error:
```
import torch.optim as optim
import torch.nn as nn


class TwoLayerNet(nn.Module):
    def __init__(self, D_in, H, D_out):
        super(TwoLayerNet, self).__init__()
        self.linear1 = nn.Linear(D_in, H)
        self.linear2 = nn.Linear(H, D_out)

    def forward(self, x):
        h_relu = self.linear1(x).clamp(min=0)
        y_pred = self.linear2(h_relu)
        return y_pred


model = TwoLayerNet(1, 1, 1)
start_lr = 10**-7
end_lr = 1


def l1(epoch): return 0.95


def l2(epoch): return 10**epoch


def l3(epoch): return math.exp(epoch)


optimizer = optim.Adam(model.parameters(), lr=start_lr)

scheduler = optim.lr_scheduler.LambdaLR(
    optimizer=optimizer, lr_lambda=l3)
```
Where `lr_lamda=l1 ` and `lr_lambda=l2` will work, but `lr_lambda=l3` produces the following error:
```
File "/usr/local/lib/python3.7/site-packages/torch/optim/lr_scheduler.py", line 189, in __init__
    super(LambdaLR, self).__init__(optimizer, last_epoch)
  File "/usr/local/lib/python3.7/site-packages/torch/optim/lr_scheduler.py", line 77, in __init__
    self.step()
  File "/usr/local/lib/python3.7/site-packages/torch/optim/lr_scheduler.py", line 150, in step
    for param_group, lr in zip(self.optimizer.param_groups, values):
UnboundLocalError: local variable 'values' referenced before assignment
```

**Comment by mattip** [comment]
This was fixed in gh-32997, which was merged after the release of 1.4.0 and is part of 1.5.0. Any exceptions in the `with` clause that defines `values` were being swallowed.

**Comment by vincentqb** [comment]
Since #32997 landed and is part of 1.5.0, I will close this issue. Please feel free to re-open if this is still happening with the latest version.
